src/test/java/com/jadajabajava/providers/HhVacancyProviderTest.java

```diff
@@ -0,0 +1,163 @@
+
+import java.io.IOException;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.junit.jupiter.api.Assertions.fail;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.argThat;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+@ExtendWith(MockitoExtension.class)
+class HhVacancyProviderTest {
+
+    @Mock
+    private HttpClient httpClient;
+
+    @Mock
+    private HttpResponse<String> httpResponseList;
+
+    @Mock
+    private HttpResponse<String> httpResponseVacancy;
+
+    @Mock
+    private ObjectMapper mapper;
+
+    @Mock
+    private MessageSource messageSource;
+
+    @Mock
+    private HhVacancyDeserializer vacancyDeserializer;
+
+    @InjectMocks
+    private HhVacancyProvider vacancyProvider;
+
+    @Captor
+    private ArgumentCaptor<HttpRequest> requestCaptor;
+
+    private static final String VACANCIES_LIST_VALID_JSON_FILE_PATH = "json/vacancies_list_valid.json";
+    private static final String VACANCIES_LIST_INVALID_JSON_FILE_PATH = "json/vacancies_list_invalid.json";
+    private static final String VACANCIES_LIST_EMPTY_JSON_FILE_PATH = "json/vacancies_list_empty.json";
+    private static final String VACANCY_VALID_JSON_FILE_PATH = "json/vacancy_valid.json";
+
+    private final ClassLoader classLoader = this.getClass().getClassLoader();
+
+    // handle invalid response
+    // handle valid response
+
+    @Test
+    void requestVacancies_buildsRequestCorrectly() {
+        String query = "queryThatProducesEmptyResponse";
+        Long area = 2019L;
+
+        try {
+            String validJson = new String(Objects.requireNonNull(classLoader.getResourceAsStream(VACANCIES_LIST_EMPTY_JSON_FILE_PATH)).readAllBytes());
+
+            when(httpResponseList.statusCode()).thenReturn(HttpStatus.OK.value());
+            when(httpResponseList.body()).thenReturn(validJson);
+            when(httpClient.send(any(), eq(HttpResponse.BodyHandlers.ofString()))).thenReturn(httpResponseList);
+
+            vacancyProvider.requestVacancies(query, List.of(area));
+
+            verify(httpClient).send(requestCaptor.capture(), any());
+        } catch (VacancyProviderException | IOException | InterruptedException e) {
+            fail("Unexpected exception: " + e.getMessage(), e);
+        }
+
+        String capturedRequestQuery = requestCaptor.getValue().uri().getQuery();
+
+        assertTrue(capturedRequestQuery.contains(query));
+        assertTrue(capturedRequestQuery.contains(String.valueOf(area)));
+    }
+
+    @Test
+    void requestVacancies_invalidResponse_handlesCorrectly() {
```

Comment:
Codacy found an issue: [The JUnit 4 test method name 'requestVacancies_invalidResponse_handlesCorrectly' doesn't match '[a-z][a-zA-Z0-9]*'](https://app.codacy.com/gh/jada-jaba-java/job-seeker-service/pullRequest?prid=5720132)